pyteal/stack_frame.py

```diff
@@ -318,46 +331,96 @@ def _best_frame_as_pyteal_frame(self) -> "PyTealFrame | None":
+            ScratchStackStore,
+            ScratchStore,
             Seq,
+            SubroutineCall,
             SubroutineDeclaration,
+            TxnaExpr,
+            UnaryExpr,
         )
+
         from pyteal.ast.frame import Proto
+        from pyteal.ast.return_ import ExitProgram
 
-        for expr in exprs:
-            e = cast(Expr, expr)
-            func(e)
+        for e in exprs:
+            if e.stack_frames.user_defined() and exit_on_user_defined:
+                continue
 
-            match e:
+            supported_type = True
+            expr = cast(Expr, e)
+            walker_args: list[Expr] = []
+            match expr:
                 case Assert():
-                    cls._walk_asts(func, *e.cond)
+                    walker_args = expr.cond
                 case BinaryExpr():
-                    cls._walk_asts(func, e.argLeft, e.argRight)
+                    walker_args = [expr.argLeft, expr.argRight]
                 case Cond():
-                    cls._walk_asts(func, *(y for x in e.args for y in x))
+                    walker_args = [y for x in expr.args for y in x]
+                case ExitProgram():
+                    walker_args = [expr.success]
                 case Proto():
-                    cls._walk_asts(func, e.mem_layout)
+                    walker_args = [expr.mem_layout]
                 case Seq():
-                    cls._walk_asts(func, *e.args)
+                    walker_args = cast(list[Expr], expr.args)
                 case Return():
-                    cls._walk_asts(func, e.value)
+                    if expr.value:
+                        walker_args = [expr.value]
                 case SubroutineDeclaration():
-                    cls._walk_asts(func, e.body)
-                case _:
-                    # TODO: implement more cases, but no need to error as this isn't used for functionality's sake.
+                    walker_args = [expr.body]
+                case ScratchStore():
+                    walker_args = [expr.value]
+                    if expr.index_expression:
+                        walker_args.append(expr.index_expression)
+                case SubroutineCall():
+                    subdef = expr.subroutine.get_declaration_by_option(False)
+                    if subdef:
+                        walker_args.append(subdef)
+                    subdef = expr.subroutine.get_declaration_by_option(True)
+                    if subdef:
+                        walker_args.append(subdef)
+                case UnaryExpr():
+                    walker_args = [expr.arg]
+                case Int(), MethodSignature(), ScratchStackStore(), TxnaExpr():
                     pass
+                case _:
+                    supported_type = False
+
+            should_stop = True
+            if supported_type or force_root_apply:
+                should_stop = func(expr)
+            if exit_when_stop_signal and should_stop:
+                return
```

Comment:
nit: should we return here or continue in the loop here?

Reply:
We want to exit the traversal under this condition. In the "should_stop" case (which is `True` for the main case of `_reframe_asts()` ) we are basically using `func == set_frames`'s output to determine whether to continue or not. An important exit condition is when `e.stack_frames == stack_frames` which alerts us that we've already seen this `Expr` as we've previously reset it's `stack_frames` and therefore avoid an infinite loop.

I'll add this explanation as a doc-string

Reply:
I see, that makes sense to me, and seems rather simpler than what I thought. I guess this captures any recursion case that will eventually reach back to the function call itself.

Reply:
Thinking more about it, I'm no longer 100% sure. I see I have 2 seperate logics. I will investigate the effects of just continuing instead of returning

Reply:
Ok. Thanks for your original suggestion. I've unified and trimmed all the exit logic thanks to the nudge. With the logic simplified, it was much easier to reason about, and now it seems clear that it's preferable to `continue` rather than `return`. That's because there is no obvious ordering for the AST that implies we should eagerly exit that traversal without visiting subsequent siblings. I think I was being overly conservative originally, trying to only explore the AST when absolutely necessary, and preferring to get a "miss" versus an over-aggressive and wrong mapping. However, this was before I implemented the skip strategies for not re-writing frames and not overwriting user-defined frames.

cf: 622bb848ebffa6b8e81dbecde58338e90ff29793

Reply:
looks good, thanks for the move!